### ISSUE TITLE
Help modifying the unicode value for Space

### ISSUE THREAD

**Issue by diepde** [googlei18n/noto-emoji#246]
Tl;dr: busy with a project for non-rooted Huawei devices to change emoji set using built-in theme manager, able to build font file and apply it successfully but there are issues with spacing (eg. thisishowitlooks). opened file in FontLabs and found the issue. Changing the unicode value for space to "200D" fixes the issue I have. Exporting the file from FontLabs presents its own issue so I want to change the unicode value to "200D" when the NotoColorEmoji file is building.

I am currently busy with a project which will allow non-rooted Huawei devices to use a different emoji set through the theme manager Huawei provides.
So far I have been able to build the NotoColorEmoji file without any issues following the instructions. I included the file in a Huawei Theme file (.hwt) and I was able to apply it successfully with some issues unfortunately. The issue I was facing was that spaces were not being inserted between words (eg. thisishowitlooks). I managed to fix this issue by loading in the file I created and a different font file which doesn't have the spacing issue into FontLabs and compared the two files. What I found was the my file assigned the space "character" the unicode value 0020 and the file known to be working assigned the space "character" the unicode value 200D. After manually changing this value in FontLabs the spacing issue was fixed but a new problem occured, I am no longer able to select skin tones for the emoji characters (the emoji is shown but when you press and hold the emoji to change skin colour you see the emoji then the skin colour next to it) and I believe this is an issue cause by FontLabs when it exports the font file for use in Android.

Thanks in advance

**Comment by dougfelt** [comment]
I'm having trouble following your description.

It sounds like in your environment, space is displaying with zero width.  It should have the standard advance width, so I'm not sure why this isn't rendering as you expect.

You are going to break things if you just go changing the font CMAP randomly like this.  200D should keep the glyph id it's assigned (probably glyph id 4), because most of the ligature entries depend on it.  0020 should map to glyph id 3 "space", that's what most systems expect.  The problem seems to be that something is ignoring or zeroing the advance width assigned to glyph id 3.  Maybe your environment requires there to be an entry in the glyph table for this id, there might not be one, I can't remember.  Or maybe FontLabs sees there is no glyph and "helpfully" zeros the advance.

**Comment by diepde** [comment]
As far as I can tell the problem is with the way Huawei handles applying the NotoColorEmoji file without root. 

When opening the NotoColorEmoji file in FontLab I can see that there are only emojis in the file and the space is assigned the unicode value "0020" which it should normally be. When you apply the file without root it replaces your font with only emojis but because there is nothing to define how the font should look in the NotoColorEmoji file it falls backs to a default font to load letters and numbers. But because the space is defined in the NotoColorEmoji file it for some reason recognizes it as zero width.

I'll try figure out how FontLab compiles its fonts so that I can compile it without the issue of not being able to change skin colour or gender on emojis

Thanks for your help

**Comment by brawer** [comment]
@diepde: If you want to tweak an existing font with minimal changes, have a look at the ttx tool in [fonttools](https://github.com/fonttools/fonttools).

**Comment by diepde** [comment]
I looked at fonttools but it didn't help with the issues I was facing. I went back to FontLab to play with settings related to building the font and was able to fix most of the issues I was facing.
